spinnaker_components/common/input_filtering.c

```diff
@@ -0,0 +1,320 @@
+  filter->step = _none_filter_step;
+}
+
+/* 1st Order Low-Pass ********************************************************/
+struct _lowpass_state
+{
+  value_t a;  // exp(-dt / tau)
+  value_t b;  // 1 - a
+};
+
+void _lowpass_filter_step(uint32_t n_dims, value_t *input,
+                          value_t *output, void *pars)
+{
+  // Cast the params
+  struct _lowpass_state *params = (struct _lowpass_state *) pars;
+
+  // Apply the filter to every dimension (realised as a Direct Form I digital
+  // filter).
+  for (uint32_t d = 0; d < n_dims; d++)
+  {
+    output[d] *= params->a;
+    output[d] += input[d] * params->b;
+  }
+}
+
+void _lowpass_filter_init(void *params, struct _if_filter *filter,
+                          uint32_t size)
+{
+  use(size);
+
+  // Copy the filter parameters into memory
+  MALLOC_OR_DIE(filter->state, sizeof(struct _lowpass_state));
+  spin1_memcpy(filter->state, params, sizeof(struct _lowpass_state));
+
+  debug(">> Lowpass filter (%k, %k)\n",
+            ((struct _lowpass_state *)filter->state)->a,
+            ((struct _lowpass_state *)filter->state)->b);
+
+  // Store a reference to the step function
+  filter->step = _lowpass_filter_step;
+}
+
+/* General purpose LTI implementation ****************************************/
+/* A Direct Form I implementation for linear digital filters.  This
+ * implementation cause unnecessary numbers of pipeline flushes, so fixed order
+ * filters can be implemented to reduce the cost of branch prediction failures.
+ */
+struct _lti_state
+{
+  uint32_t order;  // Order of the filter
+  value_t *a;   // Coefficients from the numerator
+  value_t *b;   // Coefficients from the denominator
+
+  // Previous values of the input (`xz`) and output (`yz`).  These are both 2D
+  // arrays, to access the kth last value of dimension d of x use:
+  //
+  //     xz[d*order + (n - k) % order]
+  //
+  value_t *xz;
+  value_t *yz;
+
+  // We treat the previous values as two circular buffers. After a simulation
+  // step `n` should be incremented (mod order) to rotate the ring.
+  int32_t n;
+};
+
+void _lti_filter_step(uint32_t n_dims, value_t *input,
+                      value_t *output, void *s)
+{
+  // Cast the state
+  struct _lti_state *state = (struct _lti_state *) s;
+
+  // Reference to previous values of x and y;
+  value_t *x, *y;
+
+  // Apply the filter to every dimension (realised as a Direct Form I digital
+  // filter).
+  for (uint32_t d = 0; d < n_dims; d++)
+  {
+    // Point to the correct previous x and y values.
+    x = &state->xz[d * state->order];
+    y = &state->yz[d * state->order];
+
+    // Create the new output value for this dimension
+    output[d] = 0.0k;
+
+    // Direct Form I filter
+    for (int32_t k=0; k < (int32_t) state->order; k++)
+    {
+      output[d] -= state->a[k] * y[(state->n - k - 1) % state->order];
+      output[d] += state->b[k] * x[(state->n - k - 1) % state->order];
```

Comment:
% by runtime value is going to result in a software divide. I would say the options are, store power of two number of, (potentially zeroed coefficients) and mask them or seek some sort of reason within https://github.com/SpiNNakerManchester/spinn_common/blob/master/src/circular_buffer.c